### PR TITLE
feat: support online batch and full table aggregation

### LINKED ISSUE #2499: Real-time feature calculation for full table
**Describe the feature you'd like**
```
+---+-----+---+-------+-------------------+---------+----------+--------+--------+-----+
| id|  uid|mid| cardno|         trans_time|trans_amt|trans_type|province|    city|label|
+---+-----+---+-------+-------------------+---------+----------+--------+--------+-----+
|  1|UserA| M1|1111111|2022-02-02 10:00:00|     10.0|      CASH|      SH|ShangHai| true|
|  3|UserA| M2|1111111|2022-02-03 11:00:00|     20.0|       POS|      SH|ShangHai| true|
|  4|UserA| M3|1111111|2022-02-07 11:00:00|     15.0|       POS|      SH|ShangHai| true|
|  6|UserA| M1|1111111|2022-02-07 13:00:00|     20.0|      CASH|      BJ| BeiJing| true|
|  8|UserA| M2|1111111|2022-02-08 11:00:00|     50.0|      CASH|      BJ| BeiJing| true|
|  9|UserA| M3|1111111|2022-02-08 12:00:00|    100.0|       POS|      BJ| BeiJing| true|
|  2|UserB| M3|2222222|2022-02-01 11:30:00|     10.0|       POS|      SH|ShangHai| true|
|  5|UserB| M1|2222222|2022-02-07 10:00:00|     20.0|       POS|      SH|ShangHai| true|
|  7|UserB| M2|2222222|2022-02-07 12:00:00|     30.0|       POS|      SH|ShangHai| true|
| 10|UserB| M5|2222222|2022-02-08 11:00:00|     10.0|       POS|      SH|ShangHai| true|
+---+-----+---+-------+-------------------+---------+----------+--------+--------+-----+
```
比如：查看city最近10分钟出现最多的城市
这种deployment是全表统计的，实时特征计算应当无需任何input，而当前的deployment的partition字段和请求的input（类似{"input":[{{rowdata}}],"need_schema":false}）字段是必须
请问是否考虑这种实时特征计算的支持呢

**Additional context**

### LINKED ISSUE #2618: The result of online full table aggregation is not correct and not consistent in multiple runs
**Bug Description**

image

**Expected Behavior**

**Steps to Reproduce**

1. data is generated as follows: `for i in {0..999999}; do echo "key-$((i%100)),20,22,9.2,19.3,$((1636207999998+i)),2021-01-10" >> /mnt/hdd0/zhanghao/data-100k.csv; done`
2. table creation: `CREATE TABLE demo_table2(c1 string, c2 int, c3 bigint, c4 float, c5 double, c6 timestamp, c7 date, index(key=c1, ts=c6)) options(partitionnum=8, replicanum=1);`
3. two tablets in total
4.

### LINKED ISSUE #2629: DistributedIterator mess up the key if there are multiple indexes
**Bug Description**

sql case:
```yaml
  -
    id: 3
    desc: 3 window,pk不同
    inputs:
      -
        columns : ["id int","c1 string", "c2 string", "c3 int","c4 bigint","c7 timestamp"]
        indexs: ["index1:c1:c7", "index2:c2:c7", "index3:c3:c7"]
        rows:
          - [1,"aa", "1", 20,30,1590738990000]
          - [2,"aa", "2", 20,31,1590738991000]
          - [3,"bb", "1", 20,32,1590738992000]
          - [4,"bb", "2", 20,33,1590738993000]
          - [5,"cc", "1", 21,34,1590738994000]
          - [6,"aa", "1", 21,35,1590738995000]
          - [7,"aa", "1", 21,36,1590738996000]
    sql: |
      SELECT id, c1, c2, c3,
      count(id) OVER w1 as w1_count,
      count(id) OVER w2 as w2_count,
      sum(c4) OVER w3 as w3_c4_sum
      FROM {0} WINDOW
      w1 AS (PARTITION BY c1 ORDER BY c7 ROWS BETWEEN 2 PRECEDING AND CURRENT ROW),
      w2 AS (PARTITION BY c2 ORDER BY c7 ROWS BETWEEN 3 PRECEDING AND CURRENT ROW),
      w3 AS (PARTITION BY c3 ORDER BY c7 ROWS BETWEEN 2 PRECEDING AND CURRENT ROW);
    expect:
      order: id
      columns: ["id int","c1 string", "c2 string", "c3 int","w1_count bigint","w2_count bigint", "w3_c4_sum bigint"]
      rows:
        - [1,"aa", "1", 20, 1, 1, 30]
        - [2,"aa", "2", 20, 2, 1, 61]
        - [3,"bb", "1", 20, 1, 2, 93]
        - [4,"bb", "2", 20, 2, 2, 96]
        - [5,"cc", "1", 21, 1, 3, 34]
        - [6,"aa", "1", 21, 3, 4, 69]
        - [7,"aa", "1", 21, 3, 4, 105]
```

Got unexpected result:
```
- [3, "bb", "1", 20, 1, 2, 93]
- [4, "bb", "2", 20, 2, 2, 96]
- [1, "aa", "1", 20, 1, 1, 30]
- [2, "aa", "2", 20, 2, 1, 61]
```

By turning on the debug, we see the `TabletPartition` output is:
image

### LINKED ISSUE #2631: rows with timestamp == 0 will be ignored in cluster mode
**Bug Description**
If there are rows with timestamp == 0, those rows is not read back in the cluster mode (multiple tablets).

There may be some bugs in the `DistributeWindowIterator`

**Steps to Reproduce**

1. load tables with ts == 0 data
2. run online batch window sql

### LINKED ISSUE #2633: window data is not correct in `window union` in online batch
**Bug Description**
```
  - id: 19
    desc: window union rows_range-current_row-instance_not_in_window
    inputs:
      - columns: ["id int","c1 string","c3 int","c4 bigint","c5 float","c6 double","c7 timestamp","c8 date"]
        indexs: ["index1:c1:c7"]
        rows:
          - [1,"aa",20,30,1.1,2.1,1590738993000,"2020-05-01"]
          - [4,"dd",20,33,1.4,2.4,1590738994000,"2020-05-04"]
          - [5,"ee",21,34,1.5,2.5,1590738995000,"2020-05-05"]
      - columns: ["id int","c1 string","c3 int","c4 bigint","c5 float","c6 double","c7 timestamp","c8 date"]
        indexs: ["index1:c3:c7"]
        rows:
          - [2,"bb",20,31,1.2,2.2,1590738991000,"2020-05-02"]
          - [3,"cc",20,32,1.3,2.3,1590738992000,"2020-05-03"]
    sql: |
      SELECT id, c1, c3, sum(c4) OVER w1 as w1_c4_sum FROM {0} WINDOW
      w1 AS (UNION {1} PARTITION BY {0}.c3 ORDER BY {0}.c7 ROWS_RANGE BETWEEN 2s PRECEDING AND CURRENT ROW EXCLUDE CURRENT_ROW INSTANCE_NOT_IN_WINDOW);
    expect:
      order: id
      columns: ["id int","c1 string","c3 int","w1_c4_sum bigint"]
      rows:
        - [1,"aa",20,63]
        - [4,"dd",20,32]
        - [5,"ee",21,null]
```

The case will fail in batch mode. Its output is:
```
        - [1,"aa",20,63]
        - [4,"dd",20,63]
        - [5,"ee",21,null]
```

RequestMode can pass.

### PR DESCRIPTION
**Pull request by zhanghaohit**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- enable online batch and full table aggregation
- bugfix: FullTableIterator GetValue

Fixes #2618
Fixes #2629
Fixes #2631 
Fixes #2633
Fixes #2499

solved some requirements in #2574

### CODE REVIEW

File: hybridse/src/plan/planner.cc [hunk 1]
```diff
@@ -559,20 +559,8 @@ base::Status Planner::ValidateClusterOnlineTrainingOp(node::PlanNode *node) {
         return base::Status::OK();
     }
     switch (node->type_) {
-        case node::kPlanTypeProject: {
```
Comment (aceforeverd) [inline hybridse/src/plan/planner.cc:562]: update function comment

**Comment by zhanghaohit** [comment]
- add flags to limit the total scanning data
- where condition also do full table scan
- last join?

File: src/tablet/tablet_impl.cc [hunk 2]
```diff
@@ -1036,7 +1037,8 @@ int32_t TabletImpl::ScanIndex(const ::openmldb::api::ScanRequest* request, const
             continue;
         }
         uint64_t ts = combine_it->GetTs();
-        if (ts <= et) {
+        // et is inclusive
+        if (ts < et) {
```
Comment (zhanghaohit) [inline src/tablet/tablet_impl.cc]: special handling for `et == 0` only to make it compatible.

**Review by aceforeverd — APPROVED** [review]

**Comment by zhanghaohit** [comment]
add docs about `scan_max_bytes_size` and `max_traverse_cnt`

File: src/tablet/tablet_impl.cc (hunk 2 above)
Comment (dl239) [reply src/tablet/tablet_impl.cc]: should be revert in #2669

File: src/catalog/distribute_iterator.cc [hunk 3]
```diff
@@ -330,7 +347,7 @@ void DistributeWindowIterator::Next() {
             }
             cur_pid_ = iter->first;
             uint32_t count = 0;
-            kv_it_ = iter->second->Traverse(tid_, cur_pid_, "", "", 0, FLAGS_traverse_cnt_limit, false, 0, count);
+            kv_it_ = iter->second->Traverse(tid_, cur_pid_, index_name_, "", 0, FLAGS_traverse_cnt_limit, false, 0, count);
```
Comment (dl239) [inline src/catalog/distribute_iterator.cc]: cpplint

File: src/catalog/distribute_iterator.cc [hunk 4]
```diff
@@ -280,6 +299,13 @@ DistributeWindowIterator::ItStat DistributeWindowIterator::SeekByKey(const std::
 }
 
 void DistributeWindowIterator::Next() {
+    pk_cnt_++;
+    if (pk_cnt_ >= FLAGS_max_traverse_pk_cnt) {
+        PDLOG(WARNING,
+              "DistributeWindowIterator exceed the max_traverse_pk_cnt, tid %u, cnt %u, max_traverse_pk_cnt %u",
+              pk_cnt_, FLAGS_max_traverse_pk_cnt, tid_);
```
Comment (dl239) [inline src/catalog/distribute_iterator.cc]: parameter is not match

File: src/catalog/distribute_iterator_test.cc [hunk 5]
```diff
@@ -67,7 +71,7 @@ void PutKey(const std::string& key, std::shared_ptr<openmldb::storage::Table> ta
     uint64_t now = ::baidu::common::timer::get_micros() / 1000;
```
Comment (dl239) [inline src/catalog/distribute_iterator_test.cc]: rm now

File: src/catalog/distribute_iterator_test.cc [hunk 6]
```diff
@@ -86,7 +90,7 @@ void PutKey(const std::string& key, const ::openmldb::api::TableMeta& table_meta
...
```
Comment (dl239) [inline src/catalog/distribute_iterator_test.cc]: rm now

**Review by dl239 — APPROVED** [review]

**Pull request merged**